runtime/common/src/fees.rs

```diff
@@ -0,0 +1,440 @@
+// Copyright 2022-2023 Forecasting Technologies LTD.
+// Copyright 2021-2022 Zeitgeist PM LLC.
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+//
+// This file is part of Zeitgeist.
+//
+// Zeitgeist is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the
+// Free Software Foundation, either version 3 of the License, or (at
+// your option) any later version.
+//
+// Zeitgeist is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Zeitgeist. If not, see <https://www.gnu.org/licenses/>.
+
+#[macro_export]
+macro_rules! impl_fee_types {
+    () => {
+        pub struct DealWithFees;
+
+        type NegativeImbalance = <Balances as Currency<AccountId>>::NegativeImbalance;
+        impl OnUnbalanced<NegativeImbalance> for DealWithFees {
+            fn on_unbalanceds<B>(mut fees_then_tips: impl Iterator<Item = NegativeImbalance>) {
+                if let Some(mut fees) = fees_then_tips.next() {
+                    if let Some(tips) = fees_then_tips.next() {
+                        tips.merge_into(&mut fees);
+                    }
+                    debug_assert!(
+                        FEES_AND_TIPS_TREASURY_PERCENTAGE + FEES_AND_TIPS_BURN_PERCENTAGE == 100u32
+                    );
+                    let mut split = fees
+                        .ration(FEES_AND_TIPS_TREASURY_PERCENTAGE, FEES_AND_TIPS_BURN_PERCENTAGE);
+                    Treasury::on_unbalanced(split.0);
+                }
+            }
+        }
+
+        pub struct DealWithForeignFees;
+
+        impl OnUnbalanced<CreditOf<AccountId, Tokens>> for DealWithForeignFees {
+            fn on_unbalanced(fees_and_tips: CreditOf<AccountId, Tokens>) {
+                // We have to manage the mint / burn ratio on the Zeitgeist chain,
+                // but we do not have the responsibility and necessary knowledge to
+                // manage the mint / burn ratio for any other chain.
+                // Thus we should keep 100% of the foreign tokens in the treasury.
+                // Handle the split imbalances
+                // on_unbalanced is not implemented for other currencies than the native currency
+                // https://github.com/paritytech/substrate/blob/85415fb3a452dba12ff564e6b093048eed4c5aad/frame/treasury/src/lib.rs#L618-L627
+                // https://github.com/paritytech/substrate/blob/5ea6d95309aaccfa399c5f72e5a14a4b7c6c4ca1/frame/treasury/src/lib.rs#L490
+                let _ = <Tokens as Balanced<AccountId>>::resolve(
+                    &TreasuryPalletId::get().into_account_truncating(),
+                    fees_and_tips,
+                );
```

Comment:
`debug_assert!` would be useful here to know when funds are just lost.

Reply:
true, 5c028e0583b9870141d54c8fb1fadf5a8677c202